OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs

```diff
@@ -0,0 +1,220 @@
+	{
+		public CPos GetRandomBaseCenter()
+		{
+			var randomConstructionYard = world.Actors.Where(a => a.Owner == player &&
+				Info.ConstructionYardTypes.Contains(a.Info.Name))
+				.RandomOrDefault(world.LocalRandom);
+
+			return randomConstructionYard != null ? randomConstructionYard.Location : initialBaseCenter;
+		}
+
+		readonly World world;
+		readonly Player player;
+
+		readonly Predicate<Actor> unitCannotBeOrdered;
+
+		IBotPositionsUpdated[] notifyPositionsUpdated;
+		IBotRequestUnitProduction[] requestUnitProduction;
+
+		CPos initialBaseCenter;
+		int scanInterval;
+		int ticks;
+
+		// MCVs that the bot already knows about. Any MCV not on this list needs to be given an order.
+		List<Actor> activeMCVs = new List<Actor>();
+
+		public McvManagerBotModule(Actor self, McvManagerBotModuleInfo info)
+			: base(info)
+		{
+			world = self.World;
+			player = self.Owner;
+			unitCannotBeOrdered = a => a.Owner != player || a.IsDead || !a.IsInWorld;
+		}
+
+		protected override void TraitEnabled(Actor self)
+		{
+			notifyPositionsUpdated = player.PlayerActor.TraitsImplementing<IBotPositionsUpdated>().ToArray();
+			requestUnitProduction = player.PlayerActor.TraitsImplementing<IBotRequestUnitProduction>().ToArray();
+
+			// Avoid all AIs reevaluating assignments on the same tick, randomize their initial evaluation delay.
+			scanInterval = world.LocalRandom.Next(Info.ScanForNewMcvInterval, Info.ScanForNewMcvInterval * 2);
+		}
+
+		void IBotPositionsUpdated.UpdatedBaseCenter(CPos newLocation)
+		{
+			initialBaseCenter = newLocation;
+		}
+
+		void IBotPositionsUpdated.UpdatedDefenseCenter(CPos newLocation) { }
+
+		void IBotTick.BotTick(IBot bot)
+		{
+			ticks++;
+
+			if (ticks == 1)
+				DeployMcvs(bot, false);
+
+			if (--scanInterval <= 0)
+			{
+				scanInterval = Info.ScanForNewMcvInterval;
+				DeployMcvs(bot, true);
+
+				// No construction yards - Build a new MCV
+				if (ShouldBuildMCV())
+				{
+					var unitBuilder = requestUnitProduction.FirstOrDefault(Exts.IsTraitEnabled);
+					if (unitBuilder != null)
+					{
+						var mcvInfo = AIUtils.GetInfoByCommonName(Info.McvTypes, player);
+						unitBuilder.RequestUnitProduction(bot, mcvInfo.Name);
+					}
+				}
+			}
+		}
+
+		bool ShouldBuildMCV()
+		{
+			// Only build MCV if we don't already have one in the field.
+			var allowedToBuildMCV = AIUtils.CountActorByCommonName(Info.McvTypes, player) == 0;
+			if (!allowedToBuildMCV)
+				return false;
+
+			// Build MCV if we have no construction yard, unless we have no factory (can't build it).
+			return AIUtils.CountBuildingByCommonName(Info.ConstructionYardTypes, player) == 0 &&
+				AIUtils.CountBuildingByCommonName(Info.McvFactoryTypes, player) > 0;
+		}
+
+		void DeployMcvs(IBot bot, bool chooseLocation)
+		{
+			activeMCVs.RemoveAll(unitCannotBeOrdered);
+
+			var newMCVs = world.ActorsHavingTrait<IPositionable>()
```

Comment:
Can dramatically reduce the set of actors to check by querying `ActorsHavingTrait<Transforms>` instead!